transport/src/main/java/io/netty/channel/FileChannelBasedFileRegion.java

```diff
@@ -0,0 +1,161 @@
+ * under the License.
+ */
+package io.netty.channel;
+
+import io.netty.util.IllegalReferenceCountException;
+
+import java.io.IOException;
+import java.nio.channels.FileChannel;
+import java.nio.channels.WritableByteChannel;
+
+/**
+ * Base class for {@link FileRegion} implementations which based on a {@link FileChannel}.
+ */
+public abstract class FileChannelBasedFileRegion implements SliceableFileRegion {
+
+    private long transferIndex;
+
+    @Override
+    public long transferBytesTo(WritableByteChannel target, long length) throws IOException {
+        long written = transferBytesTo(target, transferIndex, length);
+        if (written > 0) {
+            transferIndex += written;
+        }
+        return written;
+    }
+
+    @Override
+    public long transferIndex() {
+        return transferIndex;
+    }
+
+    @Override
+    public FileChannelBasedFileRegion transferIndex(long index) {
+        if (index < 0 || index > count()) {
+            throw new IndexOutOfBoundsException(
+                    String.format("transferIndex: %d (expected: 0 <= transferIndex <= count(%d))",
+                            index, count()));
+        }
+        this.transferIndex = index;
+        return this;
+    }
+
+    @Override
+    public boolean isTransferable() {
+        return transferIndex < count();
+    }
+
+    @Override
+    public long transferableBytes() {
+        return count() - transferIndex;
+    }
+
+    @Override
+    public long transferBytesTo(WritableByteChannel target, long position, long length)
+            throws IOException {
+        if (position < 0 || position > count()) {
+            throw new IllegalArgumentException("position out of range: " + position
+                    + " (expected: 0 - " + (this.count() - 1) + ')');
+        }
+        if (length < 0) {
+            throw new IllegalArgumentException("negative length " + length);
+        }
+        length = Math.min(count() - position, length);
+        if (length == 0) {
+            return 0;
+        }
+        open();
+        return channel().transferTo(position() + position, length, target);
+    }
+
+    @Override
+    public SliceableFileRegion slice(long index, long length) {
+        return new FileChannelBasedSlicedFileRegion(this, index, length);
+    }
+
+    @Override
+    public SliceableFileRegion retainedSlice(long index, long length) {
+        SliceableFileRegion sliced = slice(index, length);
+        sliced.retain();
+        return sliced;
+    }
+
+    @Override
+    public SliceableFileRegion transferSlice(long length) {
+        SliceableFileRegion sliced = slice(transferIndex(), length);
+        transferIndex += length;
+        return sliced;
+    }
+
+    @Override
+    public SliceableFileRegion transferRetainedSlice(long length) {
+        SliceableFileRegion sliced = transferSlice(length);
+        sliced.retain();
+        return sliced;
+    }
+
+    private long transferred;
```

Comment:
can you move this member variable to the top of the class declaration?

Reply:
Done.